[#9693] refactor isEditAllowed

JonasCir opened the pull request:
Fixes #9693 
Turns out everything is fine, the checks were just scattered around, I cleaned it up and now everything is just in one place.

MateStrysewske reviewed (commented):
I'm a bit doubtful about this PR, honestly. First of all, a method that's called "isEditAllowed" returning anything but a boolean is very counterintuitive to me (I know that this method existed before, but if we're doing a refactoring here, I think renaming the getEditPermissionType methods to this is the wrong way to go). Alternatives could be to only use this method when we're really interested in the ALLOWED enum value and then letting isEditAllowed return a boolean - or to keep the getEditPermissionType name and use it everywhere (and completely get rid of the isEditAllowed name).

Secondly, I don't like replacing the explicit classes with superclasses for the UUID, ARCHIVED and DELETED calls. I know that the IDE suggests those, but @StefanKock has a very good point in saying that using the explicit classes gives more clues as to what exactly we're interested in, and I think that makes a lot of sense. I've added him as another reviewer so that he can say something about it.

Bottom line: I think we should discuss in which way this PR makes sense and whether we might need to change stuff about it.

StefanKock commented on sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactJurisdictionFlagsDto.java:
```diff
@@ -20,19 +20,19 @@
 
-    private Boolean isInJurisdiction;
-    private Boolean isCaseInJurisdiction;
+    private final boolean isInJurisdiction;
```
Why this refactoring? Is it always decidable without null as third option?

JonasCir replied:
removed from PR.

StefanKock commented on sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreAdoService.java:
```diff
@@ -99,7 +98,7 @@ public Map<String, Date> calculateEndOfProcessingDate(List<String> entityuuids)
 		Expression aggregatedChangeDateExpression = addChangeDates(new AggregatedChangeDateExpressionBuilder(cb), from, true).build();
 		cq.multiselect(from.get(AbstractDomainObject.UUID), cb.max(aggregatedChangeDateExpression));
-		cq.where(from.get(ADO.UUID).in(entityuuids));
+		cq.where(from.get(AbstractDomainObject.UUID).in(entityuuids));
```
Use `CoreAdo`

JonasCir replied:
done

StefanKock reviewed (changes requested):
I totally agree with @MateStrysewske:
1. `isEditAllowed` not returning a `boolean` but a currently tri-state enum is a lot against naming conventions. I think it should be the other way around: `getEditPermissionType` is valid and `isEditAllowed` has to be removed or changed to `return getEditPermissionType == EditPermissionType.ALLOWED;` (as convenience, but is still slightly duplicate).
2. Changes of UUID / ARCHIVED constants to superclasses should always reflect the class that you expect to handle. In `AbstractCoreAdoService` changing it all to `CoreAdo` (instead of `Case`) is reasonable and clean code. But don't change it in `CampaignService` since it makes clear that it's the Campaigns attributes.

JonasCir commented:
@MateStrysewske @Totally agree.

So I updated the PR quite a bit which gives as good wins w.r.t. readability and LOC.

1. `isEditAllowed` returns a boolean as it should, making the call sites much cleaner.
2. I refactored the edit checks in S2S in such a way that they do not rely on calls into subclasses anymore
3. I introduced the `AbstractEditAllowedDetailView` which removes quite some code duplication from the UI w.r.t. edit checks

StefanKock commented on sormas-ui/src/main/java/de/symeda/sormas/ui/samples/AbstractSampleView.java:
```diff
@@ -118,15 +118,15 @@ protected void setSubComponent(DirtyStateComponent newComponent) {
 
-	protected Boolean isSampleEditAllowed() {
-		return FacadeProvider.getSampleFacade().isSampleEditAllowed(getSampleRef().getUuid());
+	protected Boolean isEditAllowed() {
```
Why keep Object Boolean here?

JonasCir replied:
changed

StefanKock commented on sormas-backend/src/main/java/de/symeda/sormas/backend/sample/SampleService.java:
```diff
@@ -1094,7 +1094,8 @@ public Predicate createDefaultFilter(CriteriaBuilder cb, Root<Sample> root) {
 
-	public Boolean isSampleEditAllowed(Sample sample) {
+
+	public Boolean isEditAllowed(Sample sample) {
```
This should not be an Object Boolean, and the current implementation does not support it. 
`inJurisdictionOrOwned(sample).getInJurisdiction()::Boolean` will run into NullPointerException.

JonasCir replied:
fixed

StefanKock commented on sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractDetailView.java:
```diff
@@ -115,4 +115,5 @@ protected R getReference() {
 	 *            The URL parameters String
 	 */
 	protected abstract void initView(String params);
+
```
Why empty line here? It's against code convention

JonasCir replied:
fixed

StefanKock commented on sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractEditAllowedDetailView.java:
```diff
@@ -0,0 +1,34 @@
+
+	private CoreFacade coreFacade;
+
+	protected AbstractEditAllowedDetailView(String viewName, CoreFacade coreFacade) {
```
This pattern caches the Facade instance which is discouraged because it does not refresh when the sormas-ear is redeployed.

JonasCir replied:
fixed

StefanKock reviewed (changes requested):
The approach with usage of `isEditAllowed` and `getEditPermissionType` looks now way cleaner. But some new remarks.

JonasCir commented on sormas-backend/src/main/java/de/symeda/sormas/backend/common/AbstractCoreFacadeEjb.java:
```diff
@@ -259,8 +259,13 @@ public Date calculateEndOfProcessingDate(String entityUuid) {
-		ADO ado = service.getByUuid(uuid);
-		return service.isEditAllowed(ado);
+	@Override
+	public EditPermissionType getEditPermissionType(String uuid) {
```
Good point, thanks. I think this is a fundamental utility method which should be callable unconditionally.

MateStrysewske commented:
@JonasCir Code looks good to me, can you resolve the merge conflict please?
@StefanKock Can you do another review?

JonasCir commented:
conflict resolved

MateStrysewske commented:
I'll have to postpone testing this because it's blocked by https://github.com/hzi-braunschweig/SORMAS-Project/issues/8541.

JonasCir commented:
@MateStrysewske Thx, can we merge or should we wait for a review from @StefanKock ? :)

MateStrysewske commented:
Wait for his approval please :)

StefanKock reviewed (approved):
Thx for waiting ;-) I have no objections.

JonasCir commented:
All good, thank you both! :rocket:
